### ISSUE TITLE
NaNs in user-generated .pegs4dat file

### ISSUE THREAD

**Issue by ojalaj** [nrc-cnrc/EGSnrc#85]
Hello.

With following `.pegs4inp` file I get NaNs in the generated `.pegs4dat` file:
```
MIXT
 &INP NE=5,RHO=7.8700E+00,RHOZ=0.0045,0.9875,0.0071,0.0004,0.0005,IRAYL=1, EPSTFL=0 &END 
AISI1045
C  FE MN P  S
ENER
 &INP AE=0.521,UE=55.511,AP=0.01,UP=55 &END
PWLF
 &INP  &END
DECK
 &INP  &END
```
Is this normal? At least it crashes the simulations. I've never seen this before, when using pegs4.

PS: Issue #67 also has an effect, when using from command line (at least I wasn't able to succeed in generating .pegs4dat with density corrections). I haven't tested yet if fix #80 help, but will do it, when merged to `develop` branch.

**Comment by ftessier** [comment]
NaNs in pegs4 files can arise due to the `-ffast-math` compiler flag when compiling the `pegs4` executable. Go inside your `$EGS_CONFIG` file and remove `-ffast-math` from the Fortran compiler flags, then go in `$HEN_HOUSE/pegs4` to compile using `make clean` and `make`.

**Comment by ojalaj** [comment]
I tried and NaNs are no longer present - thank you.

The reason why I haven't seen these before might be that I've previously generated all the pegs4 data in a Windows-based EGS/BEAM distribution, so this flag might not be there and/or something else has prevented NaNs from being generated. Since -ffast-math flag was suggest during configuration by default, could this kind of issue be worth mentioning in the documentation?

**Comment by ftessier** [comment]
Good point. The flag is not suggested when using `$HEN_HOUSE/scripts/configure`, but I see it in the GUI configuration tool; @mainegra can you remove the `-ffast-math` default for the Fortran compiler in [`egs_tools.cpp`](https://github.com/nrc-cnrc/EGSnrc/blob/4fb7c06fb8b262834132d2860a8b015e03d7e5b3/HEN_HOUSE/gui/egs_configure/egs_tools.cpp)?

**Comment by ojalaj** [comment]
Some further suggestion/question - if you are aware some kind of behaviour, which is due to using/not using some options/flags, for compilers that you suggest to be used as default, could you mention them in the documentation and/or in the configuration phase? I know that there are many different compilers available and you can't support them all 100%, but at least for the recommended ones. E.g.following a suggestion by one of my colleagues, I'm using '-O3 -march=native -ffast-math -funroll-loops' for the gfortran, but honestly I'm not sure could it generate some further problems (?).

**Comment by ftessier** [comment]
Ideally yes, but there such a long list of compiler flags, this is not practical. Our approach is to propose a very minimal set of compiler flags that are expected to work across the board for most users, and let more advanced users tweak their compilers as they see fit. This is standard practice.

**Comment by ojalaj** [comment]
OK - I fully understand and agree. Thank you!

**Comment by mainegra** [comment]
This issue might be caused by a specific version of the GNU Fortran compiler which may show in one OS and not in the other. Up until recently `-ffast-math` did not cause any problems. It was good to have as it did produce faster code. We could be cautious and stop suggesting optimization options leaving it up to the users. The downside of this could be having users never looking into exploiting the compiler's full potential.

**Comment by ftessier** [comment]
Agreed. This particular problem with `-ffast-math` and pegs4 has come up a number of times recently, so I suggest we remove it from the suggested default in the configuration tool. Ideally we could figure out why it causes NaNs in the first place and fix pegs4 accordingly. Are you up for this @blakewalters?

**Comment by mainegra** [comment]
Perhaps is the compiler that needs fixing? In which case there is nothing that can be done in pegs4. But maybe worth taking a peek.

**Comment by blakewalters** [comment]
I can take a look at it.

**Comment by ftessier** [comment]
You can try expanding the individual options invoked by `-ffast-math` to find the culprit, which might give a clue as to what part of the code is failing under non-strict IEEE compliance.

https://gcc.gnu.org/onlinedocs/gcc-5.3.0/gcc/Optimize-Options.html#Optimize-Options
```
-ffast-math     Sets the options -fno-math-errno, -funsafe-math-optimizations, -ffinite-math-only, -fno-rounding-math, -fno-signaling-nans and -fcx-limited-range.
```
My guess is `-ffinite-math-only`; bets are open!

**Comment by blakewalters** [comment]
Just back from taking a look in pegs4.mortran.  I've found that the NAN values for cross-sections are traceable to an Infinite value of the parameter, VPLASM=SQRT(EDEN*R0*C**2/PI), which, in turn results in +/- Infinite values of the parameters, CBAR, and AFACT.  At this point, I'm not exactly sure what the physical meanings of VPLASM, CBAR, AFACT are.  Numerically, however, when -ffast-math is turned on, the argument, EDEN*R0*C**2/PI=Infinity (5.6e32 without -ffast-math).  The problem can be solved by dividing EDEN by PI first in the definition of VPLASM:  VPLASM=SQRT(EDEN/PI*R0*C**2).

**Comment by blakewalters** [comment]
I mean: VPLASM=SQRT(EDEN*R0*C**2/PI) which should be recast as VPLASM=SQRT(EDEN/PI*R0*C**2).

**Comment by ftessier** [comment]
Awesome!!

Thanks!

Frédéric.

**Comment by lichuanlong** [comment]
gfortran5.3 has no problum. But gfortrn4.8 compiling pegs4.exe has NaNs.

**Comment by ftessier** [comment]
@blakewalters, can you fix this in pegsless mode as well? @mainegra reports having the same issue with pegsless, when compiling with optimization turned on.

**Comment by ftessier** [comment]
Fixed in 1de38ba

**Comment by ftessier** [comment]
Fixed in 20ff345
